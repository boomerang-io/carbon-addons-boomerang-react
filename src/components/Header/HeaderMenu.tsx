import React from "react";
import { prefix } from "../../internal/settings";

type Props = {
  ["aria-labelledby"]: string;
  children?: any;
  id: string;
};

function HeaderMenu({  children, id, ...rest }: Props) {
  return (
    <ul className={`${prefix}--bmrg-header-drop-down`} id={id} role="menu" {...rest}>
      {children}
    </ul>
  );
}

export default HeaderMenu;
