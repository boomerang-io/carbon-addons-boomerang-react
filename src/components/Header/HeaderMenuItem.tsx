import React from "react";
import { HeaderMenuItem as CarbonHeaderMenuItem } from "@carbon/react";
import Avatar from "../Avatar";
import { ArrowRight, Launch } from "@carbon/react/icons";
import { prefix } from "../../internal/settings";

type Shared = {
  className?: string;
  children?: React.ReactNode;
  disabled?: boolean;
  element?: React.FC<any>;
  icon?: React.ReactNode;
  style?: React.CSSProperties;
  variant?: "danger" | "default";
};

type Props =
  | (Shared & {
      href: string;
      kind: "app" | "internal" | "external";
      onClick?: () => void;
      text: string;
      type: "link";
    })
  | (Shared & {
      onClick: () => void;
      text: string;
      type: "button";
    })
  | (Shared & {
      onClick: () => void;
      src: string;
      type: "user";
      userName?: string;
    });

function HeaderMenuItem(props: Props, ref: React.ForwardedRef<HTMLLinkElement>) {
  const { type, icon, onClick, variant = "default", ...rest } = props;

  if (props.type === "button")
    return (
        <CarbonHeaderMenuItem
          // eslint-disable-next-line no-script-url
          href={"javascript:void(0)"}
          onClick={onClick}
          role="menuitem"
          ref={ref}
          {...rest}
        >
          <div className={`${prefix}--bmrg-header-menu-item__content ${variant === "danger" ? "--danger" : ""}`}>
            <span className={`${prefix}--bmrg-header-menu-item__text`}>
              {icon}
              {props.text}
            </span>
          </div>
        </CarbonHeaderMenuItem>
    );

  if (props.type === "link") {
    const externalProps = props.kind === "external" ? { target: "_blank", rel: "noopener noreferrer" } : {};
    let linkTypeIcon;
    switch (props.kind) {
      case "external":
        linkTypeIcon = <Launch aria-label="Opens in new tab" />;
        break;
      case "internal":
        linkTypeIcon = <ArrowRight aria-label="Opens in same platform" />;
        break;
      case "app":
      default:
      // no-op
    }
    return (
      <CarbonHeaderMenuItem href={props.href} {...externalProps} onClick={onClick} role="menuitem" ref={ref}>
        <div className={`${prefix}--bmrg-header-menu-item__content ${variant === "danger" ? "--danger" : ""}`}>
          <span className={`${prefix}--bmrg-header-menu-item__text`}>
            {icon}
            {props.text}
          </span>
          {linkTypeIcon}
        </div>
      </CarbonHeaderMenuItem>
    );
  }

  if (props.type === "user") {
    return (
      // eslint-disable-next-line no-script-url
      <CarbonHeaderMenuItem href={"javascript:void(0)"} onClick={onClick} role="menuitem" ref={ref}>
        <div className={`${prefix}--bmrg-header-menu-user`}>
          <Avatar size="medium" src={props.src} userName={props.userName} />
          <p className={`${prefix}--bmrg-header-menu-user__name`}> {props.userName ? props.userName : ""} </p>
        </div>
      </CarbonHeaderMenuItem>
    );
  }

  return null;
}

export default React.forwardRef(HeaderMenuItem);
