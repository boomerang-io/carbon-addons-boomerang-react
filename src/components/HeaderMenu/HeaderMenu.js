import React from 'react';
import PropTypes from 'prop-types';
import { settings } from 'carbon-components';

const { prefix } = settings;

function HeaderMenu({ children }) {
  return (
    <div className={`${prefix}--bmrg-header-drop-down`}>
      <ul className={`${prefix}--bmrg-header-drop-down__items`}>
        {React.Children.map(children, (child) => (
          <li className={`${prefix}--bmrg-header-drop-down__item`}>{child}</li>
        ))}
      </ul>
    </div>
  );
}

HeaderMenu.propTypes = {
  children: PropTypes.any,
};

export default HeaderMenu;
