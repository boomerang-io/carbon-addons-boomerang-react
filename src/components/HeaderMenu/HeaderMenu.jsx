
import PropTypes from "prop-types";
import { prefix } from "../../internal/settings";
import FocusTrap from "focus-trap-react";



function HeaderMenu({ children }) {
  return (
    <FocusTrap active focusTrapOptions={{ allowOutsideClick: true }}>
      <div className={`${prefix}--bmrg-header-drop-down`}>
        <ul className={`${prefix}--bmrg-header-drop-down__items`} role="menu">
          {React.Children.map(children, (child) => (
            <li className={`${prefix}--bmrg-header-drop-down__item`} aria-label={`${child.key} button`} role="menuitem">
              {child}
            </li>
          ))}
        </ul>
      </div>
    </FocusTrap>
  );
}

HeaderMenu.propTypes = {
  children: PropTypes.any,
};

export default HeaderMenu;
