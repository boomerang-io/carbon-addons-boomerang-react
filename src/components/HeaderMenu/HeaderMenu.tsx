import React from "react";
import { prefix } from "../../internal/settings";
import FocusTrap from "focus-trap-react";

type Props = {
  children?: any;
  id: string;
};

function HeaderMenu({ children, id }: Props) {
  return (
    <FocusTrap active focusTrapOptions={{ allowOutsideClick: true, returnFocusOnDeactivate: false }}>
      <div className={`${prefix}--bmrg-header-drop-down`}>
        <ul className={`${prefix}--bmrg-header-drop-down__items`} role="menu" id={id}>
          {React.Children.map(children, (child) => (
            <li className={`${prefix}--bmrg-header-drop-down__item`} aria-label={`${child.key} button`} role="menuitem">
              {child}
            </li>
          ))}
        </ul>
      </div>
    </FocusTrap>
  );
}

export default HeaderMenu;
