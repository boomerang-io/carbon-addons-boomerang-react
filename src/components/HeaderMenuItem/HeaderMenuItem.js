import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { settings } from 'carbon-components';
import {
  Chat16,
  Debug20,
  Group20,
  Information16,
  Launch16,
  Locked16,
  Power16,
  Workspace20,
} from '@carbon/icons-react';

import HeaderMenuModalWrapper from '../../internal/HeaderMenuModalWrapper';

const { prefix } = settings;

const iconClassName = `${prefix}--bmrg-header-menu-item__img`;
const iconFill = '#FBFCFC';

const iconMapping = {
  workspace: <Workspace20 fill={iconFill} className={iconClassName} />,
  group: <Group20 fill={iconFill} className={iconClassName} />,
  chat: <Chat16 fill={iconFill} className={iconClassName} />,
  debug: <Debug20 fill={iconFill} className={iconClassName} />,
  power: <Power16 fill={iconFill} className={iconClassName} />,
  information: <Information16 fill={iconFill} className={iconClassName} />,
  locked: <Locked16 fill={iconFill} className={iconClassName} />,
  launch: <Launch16 fill={iconFill} className={iconClassName} />,
};

HeaderMenuItem.propTypes = {
  /**
   * alt text for carbon icon
   */
  altIconText: PropTypes.string,
  /**
   * override default className
   */
  className: PropTypes.string,
  /**
   * user specifies what to render when HeaderMenuItem is clicked
   */
  children: PropTypes.any,
  forwardRef: PropTypes.object,
  /**
   * carbon name for icon to be rendered (internally mapped)
   */
  iconName: PropTypes.oneOf([
    'workspace',
    'group',
    'chat',
    'debug',
    'power',
    'information',
    'locked',
    'launch',
  ]),
  /**
   *
   */
  style: PropTypes.object,
  /**
   * text to be displayed next to the icon
   */
  text: PropTypes.string.isRequired,

  /**
   * props to modalWrapper
   */
  disabled: PropTypes.bool,
};

HeaderMenuItem.defaultProps = {
  iconName: 'launch',
};

function HeaderMenuItem({ children, className, forwardRef, iconName, style, text, ...rest }) {
  const wrapperClassNames = cx(`${prefix}--bmrg-header-menu-item-wrapper`, {
    [className]: !!className,
  });

  return (
    <div className={wrapperClassNames} style={style} role="presentation" ref={forwardRef}>
      <HeaderMenuModalWrapper
        buttonTriggerClassName={`${prefix}--bmrg-header-menu-item`}
        buttonTriggerText={
          <div className={`${prefix}--bmrg-header-menu-item__content`}>
            <div className={`${prefix}--bmrg-header-menu-item__img`}>{iconMapping[iconName]}</div>
            <span className={`${prefix}--bmrg-header-menu-item__text`}>{text}</span>
          </div>
        }
        {...rest}
      >
        {children}
      </HeaderMenuModalWrapper>
    </div>
  );
}

export default (() => {
  const forwardRef = (props, ref) => <HeaderMenuItem {...props} forwardRef={ref} />;
  forwardRef.displayName = 'HeaderMenuItem';
  return React.forwardRef(forwardRef);
})();
