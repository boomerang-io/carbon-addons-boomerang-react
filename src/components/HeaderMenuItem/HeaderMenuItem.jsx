import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { prefix } from "../../internal/settings";
import {
  Chat,
  Debug,
  Group,
  Idea, 
  Information,
  Launch,
  Locked,
  Power,
  Workspace,
} from "@carbon/react/icons";

import HeaderMenuModalWrapper from "../../internal/HeaderMenuModalWrapper";



const iconClassName = `${prefix}--bmrg-header-menu-item__img`;
const iconFill = "#FBFCFC";

const iconMapping = {
  workspace: <Workspace size={16} fill={iconFill} className={iconClassName} />,
  group: <Group size={16} fill={iconFill} className={iconClassName} />,
  chat: <Chat size={16} fill={iconFill} className={iconClassName} />,
  debug: <Debug size={16} fill={iconFill} className={iconClassName} />,
  power: <Power size={16} fill={iconFill} className={iconClassName} />,
  information: <Information size={16} fill={iconFill} className={iconClassName} />,
  locked: <Locked size={16} fill={iconFill} className={iconClassName} />,
  launch: <Launch size={16} fill={iconFill} className={iconClassName} />,
  idea: <Idea size={16}  fill={iconFill} className={iconClassName} />,
};

HeaderMenuItem.propTypes = {
  /**
   * alt text for carbon icon
   */
  altIconText: PropTypes.string,
  /**
   * override default className
   */
  className: PropTypes.string,
  /**
   * user specifies what to render when HeaderMenuItem is clicked
   */
  children: PropTypes.any,
  forwardRef: PropTypes.object,
  /**
   * carbon name for icon to be rendered (internally mapped)
   */
  iconName: PropTypes.oneOf([
    "workspace",
    "group",
    "chat",
    "debug",
    "power",
    "information",
    "locked",
    "launch",
    "idea",
  ]),
  /**
   *
   */
  style: PropTypes.object,
  /**
   * text to be displayed next to the icon
   */
  text: PropTypes.string.isRequired,

  /**
   * props to modalWrapper
   */
  disabled: PropTypes.bool,
};

HeaderMenuItem.defaultProps = {
  iconName: "launch",
};

function HeaderMenuItem({
  children,
  className,
  forwardRef,
  iconName,
  preventCloseOnClickOutside,
  style,
  text,
  ...rest
}) {
  const wrapperClassNames = cx(`${prefix}--bmrg-header-menu-item-wrapper`, {
    [className]: !!className,
  });

  return (
    <div className={wrapperClassNames} style={style} role="presentation" ref={forwardRef}>
      <HeaderMenuModalWrapper
        preventCloseOnClickOutside={preventCloseOnClickOutside}
        buttonTriggerClassName={`${prefix}--bmrg-header-menu-item`}
        buttonTriggerText={
          <div className={`${prefix}--bmrg-header-menu-item__content`}>
            <div className={`${prefix}--bmrg-header-menu-item__img`}>{iconMapping[iconName]}</div>
            <span className={`${prefix}--bmrg-header-menu-item__text`}>{text}</span>
          </div>
        }
        {...rest}
      >
        {children}
      </HeaderMenuModalWrapper>
    </div>
  );
}

export default (() => {
  const forwardRef = (props, ref) => <HeaderMenuItem {...props} forwardRef={ref} />;
  forwardRef.displayName = "HeaderMenuItem";
  return React.forwardRef(forwardRef);
})();
