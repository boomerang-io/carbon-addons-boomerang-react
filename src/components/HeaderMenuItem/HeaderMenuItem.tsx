import React from "react";
import { prefix } from "../../internal/settings";
import { HeaderMenuItem } from "@carbon/react";
import { ArrowRight, Launch } from "@carbon/react/icons";

type Shared = {
  className?: string;
  children?: React.ReactNode;
  disabled?: boolean;
  icon: React.ReactNode;
  style?: React.CSSProperties;
  text: string;
  variant?: "danger" | "default";
};

type Props =
  | (Shared & {
      external?: boolean;
      href: string;
      kind: "link";
      onClick?: () => void;
    })
  | (Shared & {
      onClick: () => void;
      kind: "button";
    });

function BmrgHeaderMenuItem(props: Props) {
  const { kind, text, icon, onClick, variant = "default", ...rest } = props;
  if (props.kind === "button")
    return (
      <HeaderMenuItem
        aria-label={`Link for ${text}`}
        // eslint-disable-next-line no-script-url
        href={"javascript:void(0)"}
        onClick={onClick}
        {...rest}
      >
        <div className={`${prefix}--bmrg-header-menu-item__content ${variant === "danger" ? "--danger" : ""}`}>
          <span className={`${prefix}--bmrg-header-menu-item__text`}>
            {icon}
            {text}
          </span>
        </div>
      </HeaderMenuItem>
    );

  if (props.kind === "link") {
    const externalProps = props.external ? { target: "_blank", rel: "noopener noreferrer" } : {};
    return (
      <HeaderMenuItem aria-label={`Button for ${text}`} href={props.href} {...externalProps} onClick={onClick}>
        <div className={`${prefix}--bmrg-header-menu-item__content ${variant === "danger" ? "--danger" : ""}`}>
          <span className={`${prefix}--bmrg-header-menu-item__text`}>
            {icon}
            {text}
          </span>
          {props.external ? <Launch /> : <ArrowRight />}
        </div>
      </HeaderMenuItem>
    );
  }

  return null;
}

export default BmrgHeaderMenuItem;
