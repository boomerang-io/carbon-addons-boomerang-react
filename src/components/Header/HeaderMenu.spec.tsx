import React from "react";
import { expect, test } from "vitest";
import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import HeaderMenu from "./HeaderMenu";
import HeaderMenuItem from "./HeaderMenuItem";

describe("Feedback", () => {
  test("snapshot", async () => {
    const { baseElement, container } = render(
      <HeaderMenu id="header-menu">
        <HeaderMenuItem type="link" kind="external" href="https://useboomerang.io" text="Use Boomerang" />
        <HeaderMenuItem type="link" kind="internal" href="https://useboomerang.io" text="Use Boomerang" />
        <HeaderMenuItem type="link" kind="app" href="https://useboomerang.io" text="Use Boomerang" />
        <HeaderMenuItem type="button" onClick={() => void 0} text="Show Boomerang" />
        <HeaderMenuItem
          type="user"
          onClick={() => void 0}
          userName="Boomerang user"
          src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50"
        />
      </HeaderMenu>
    );
    expect(baseElement).toMatchSnapshot();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  
  test("a11y", async () => {
    const { container } = render(
      <HeaderMenu id="header-menu">
        <HeaderMenuItem type="link" kind="external" href="https://useboomerang.io" text="Use Boomerang" />
        <HeaderMenuItem type="link" kind="internal" href="https://useboomerang.io" text="Use Boomerang" />
        <HeaderMenuItem type="link" kind="app" href="https://useboomerang.io" text="Use Boomerang" />
        <HeaderMenuItem type="button" onClick={() => void 0} text="Show Boomerang" />
        <HeaderMenuItem
          type="user"
          onClick={() => void 0}
          userName="Boomerang user"
          src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50"
        />
      </HeaderMenu>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
