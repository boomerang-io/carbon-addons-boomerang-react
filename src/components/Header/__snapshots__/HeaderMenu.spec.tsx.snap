// Vitest Snapshot v1

exports[`Feedback > snapshot 1`] = `
<body>
  <div>
    <ul
      aria-labelledby="menu"
      class="cds--bmrg-header-drop-down"
      id="header-menu"
      role="menu"
    >
      <li
        role="menuitem"
      >
        <a
          class="cds--header__menu-item"
          href="https://useboomerang.io"
          rel="noopener noreferrer"
          tabindex="0"
          target="_blank"
        >
          <span
            class="cds--text-truncate--end"
          >
            <div
              class="cds--bmrg-header-menu-item__content"
            >
              <span
                class="cds--bmrg-header-menu-item__text"
              >
                Use Boomerang
              </span>
              <svg
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                title="Opens link in new tab"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13,14H3c-0.6,0-1-0.4-1-1V3c0-0.6,0.4-1,1-1h5v1H3v10h10V8h1v5C14,13.6,13.6,14,13,14z"
                />
                <path
                  d="M10 1L10 2 13.3 2 9 6.3 9.7 7 14 2.7 14 6 15 6 15 1z"
                />
              </svg>
            </div>
          </span>
        </a>
      </li>
      <li
        role="menuitem"
      >
        <a
          class="cds--header__menu-item"
          href="https://useboomerang.io"
          tabindex="0"
        >
          <span
            class="cds--text-truncate--end"
          >
            <div
              class="cds--bmrg-header-menu-item__content"
            >
              <span
                class="cds--bmrg-header-menu-item__text"
              >
                Use Boomerang
              </span>
              <svg
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                title="Opens link within same platform"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.3 3.7L13.1 7.5 1 7.5 1 8.5 13.1 8.5 9.3 12.3 10 13 15 8 10 3z"
                />
              </svg>
            </div>
          </span>
        </a>
      </li>
      <li
        role="menuitem"
      >
        <a
          class="cds--header__menu-item"
          href="https://useboomerang.io"
          tabindex="0"
        >
          <span
            class="cds--text-truncate--end"
          >
            <div
              class="cds--bmrg-header-menu-item__content"
            >
              <span
                class="cds--bmrg-header-menu-item__text"
              >
                Use Boomerang
              </span>
            </div>
          </span>
        </a>
      </li>
      <li
        role="menuitem"
      >
        <a
          class="cds--header__menu-item"
          href="javascript:void(0)"
          tabindex="0"
        >
          <span
            class="cds--text-truncate--end"
          >
            <div
              class="cds--bmrg-header-menu-item__content "
            >
              <span
                class="cds--bmrg-header-menu-item__text"
              >
                Show Boomerang
              </span>
            </div>
          </span>
        </a>
      </li>
      <li
        role="menuitem"
      >
        <a
          class="cds--header__menu-item"
          href="javascript:void(0)"
          tabindex="0"
        >
          <span
            class="cds--text-truncate--end"
          >
            <div
              class="cds--bmrg-header-menu-user"
            >
              <img
                alt="Avatar for Boomerang user"
                class="cds--bmrg-avatar --medium"
                src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50"
                title="Avatar for Boomerang user"
              />
              <p
                class="cds--bmrg-header-menu-user__name"
              >
                 
                Boomerang user
                 
              </p>
            </div>
          </span>
        </a>
      </li>
    </ul>
  </div>
</body>
`;
